File: 자료구조/Sort/quickSort.java
public class quickSort { // 이런 게 있구나,, 공부하자
	
	public static void sort(int[] a) {
		m_pivot_sort(a, 0, a.length - 1); // 중간 pivot
	}
	

	private static void m_pivot_sort(int[] a, int lo, int hi) {

		if(lo >= hi) {
			return;
		}
		
		/*
		 * 피벗을 기준으로 요소들이 왼쪽과 오른쪽으로 약하게 정렬 된 상태로
		 * 만들어 준 뒤, 최종적으로 pivot의 위치를 얻는다.
		 * 
		 * 그리고나서 해당 피벗을 기준으로 왼쪽 부분리스트와 오른쪽 부분리스트로 나누어
		 * 분할 정복을 해준다.
		 */
        
		int pivot = partition(a, lo, hi);	
		
		m_pivot_sort(a, lo, pivot-1); // lo~pivot-1
		m_pivot_sort(a, pivot, hi); // pivot~hi
	}
	
	
	
	private static int partition(int[] arr, int low, int high) {
        int pivot = arr[(low + high) / 2]; // 중간 pivot
        while (low <= high) { // 교차할 때까지 반복
            while (arr[low] < pivot) low++; // 왼쪽에서 pivot보다 큰 수 찾거나 pivot까지 오는 것 기다림
            while (arr[high] > pivot) high--;
            if (low <= high) { // 교차해 지나치지 않았으면
                swap(arr, low, high); // 약한 정렬
                low++;
                high--;
            }
        }
        return low; // 교차해 지나쳤다면 low를 다음 pivot으로 삼음
    }
	
	
	
	private static void swap(int[] a, int i, int j) {
		int temp = a[i];
		a[i] = a[j];
		a[j] = temp;
	}
	
}

File: 자료구조/Sort/quickSort.py
# 중간 피벗 방식
# pivot기준 작은 원소->lesser_arr에
# pivot기준 큰 원소->greater_arr에
def quick_sort(arr):
    if len(arr) <= 1:
        return arr
    pivot = arr[len(arr) // 2]
    lesser_arr, equal_arr, greater_arr = [], [], []
    for num in arr:
        if num < pivot:
            lesser_arr.append(num)
        elif num > pivot:
            greater_arr.append(num)
        else:
            equal_arr.append(num)
    return quick_sort(lesser_arr) + equal_arr + quick_sort(greater_arr)

# 최적화. in-place 정렬
def quick_sort(arr):
    def sort(low, high):
        if high <= low:
            return

        mid = partition(low, high)
        sort(low, mid-1)
        sort(mid, high) # pivot을 rightList에 포함시킴

    # pivot에 따라 왼쪽에 작은 값, 오른쪽에 큰 값 배치, 다음 pivot=low리턴
    def partition(low, high):
        pivot = arr[(low + high) // 2] # 중간 pivot
        while low <= high: # 두 포인터 교차할 때까지 반복-><-
            while arr[low] < pivot: # pivot보다 큰 값 찾거나, pivot 위치에 올 때까지
                low += 1
            while arr[high] > pivot:
                high -= 1
            if low <= high: # 아직 교차해 지나치지 않았으면 swap해서 약한 정렬
                arr[low], arr[high] = arr[high], arr[low]
                low, high = low + 1, high - 1
        return low # 교차해 지나쳤다면 다음 pivot이 될 low 리턴
    return sort(0, len(arr)-1)